db/types/custom/underlying_type.py

```diff
@@ -0,0 +1,29 @@
+from sqlalchemy import cast
+
+
+class HasUnderlyingType:
+    """
+    We apply this mixin and its `underlying_type` property to a custom SQLAlchemy type to designate
+    what its underlying SQLAlchemy type is. We do this, because our custom types misbehave in some
+    cases (e.g. array aggregation), and so, in those cases, we cast them to what we consider the
+    underlying type. Can also be thought of as the fallback type. For example, for URIs we fallback
+    to TEXT, and for Money we fallback to NUMERIC.
+
+    Note, doesn't use ABC, because its metaclasses clash with SQLAlchemy's metaclasses. Workaround
+    possible, but seemed burdensome.
+    """
+
+    @property
+    def underlying_type(self):
```

Comment:
This property will behave differently based on whether the inheriting class is instantiated or not (if it doesn't define its own `underlying_type`); this is undesirable for something that's overridden as a class attribute in inheriting classes when (properly) used. E.g.:
```python
class MyType(UserDefinedType, HasUnderlyingType):
    # Forgot underlying_type
    
    def get_col_spec(self, **_):
        # This results in the type name being upper case when viewed.
        # Actual usage in the DB is case-insensitive.
        return DB_TYPE.upper()


MyType().underlying_type  # NoneType
MyType.underlying_type  # reference to property function
```

Reply:
Note, I played around with this a bit, and it seems that creating an abstract property (like I did) is the only way to enforce that a concrete attribute is implemented on a concrete subclass. Using `ABC` doesn't change that: it only helps enforce it. Maybe I missed something, but if I'm right the scenario you posted cannot be solved if you want an abstract class or instance property.